lib/compress/zstd_lazy.c

```diff
@@ -973,37 +1141,55 @@ ZSTD_compressBlock_lazy_generic(
         }
 
         if (matchLength < 4) {
-            ip += ((ip-anchor) >> kSearchStrength) + 1;   /* jump faster over incompressible sections */
+            size_t const skip = ((ip-anchor) >> kSearchStrength) + 1;   /* jump faster over incompressible sections */
+            ip += skip;
+            // if (skip > 1) {
+            //     nextRep32 = ZS_Repcode_gather32(reps, ip + 1);
+            // }
             continue;
         }
 
         /* let's try to find a better solution */
         if (depth>=1)
         while (ip<ilimit) {
             ip ++;
-            if ( (dictMode == ZSTD_noDict)
-              && (offset) && ((offset_1>0) & (MEM_read32(ip) == MEM_read32(ip - offset_1)))) {
-                size_t const mlRep = ZSTD_count(ip+4, ip+4-offset_1, iend) + 4;
-                int const gain2 = (int)(mlRep * 3);
-                int const gain1 = (int)(matchLength*3 - ZSTD_highbit32((U32)offset+1) + 1);
-                if ((mlRep >= 4) && (gain2 > gain1))
-                    matchLength = mlRep, offset = 0, start = ip;
-            }
-            if (isDxS) {
-                const U32 repIndex = (U32)(ip - base) - offset_1;
-                const BYTE* repMatch = repIndex < prefixLowestIndex ?
-                               dictBase + (repIndex - dictIndexDelta) :
-                               base + repIndex;
-                if (((U32)((prefixLowestIndex-1) - repIndex) >= 3 /* intentional underflow */)
-                    && (MEM_read32(repMatch) == MEM_read32(ip)) ) {
-                    const BYTE* repMatchEnd = repIndex < prefixLowestIndex ? dictEnd : iend;
-                    size_t const mlRep = ZSTD_count_2segments(ip+4, repMatch+4, iend, repMatchEnd, prefixLowest) + 4;
+            {
+                ZS_Vec128 const rep32 = ZS_Repcode_gather32(reps, ip);
+                BYTE const* const rip = ip;
+                U32 const repcode = ZS_Repcode_search(reps, rep32, rip);
+                if (repcode != 0) {
+                    BYTE const* repMatch = rip + ZS_Repcode_get(reps, repcode);
+                    size_t const mlRep = ZSTD_count(rip+4, repMatch + 4, iend) + 4;
                     int const gain2 = (int)(mlRep * 3);
```

Comment:
minor FYI : 
this formula, which makes a crude selection of current `repcode` versus previous "full offset" match based on a fixed cost evaluation, is expecting `mlCode == 0`. 
Now, if `mlCode` can also be `1` or `2`, this formula might need an adjustment.